partiql-parser/src/location.rs

```diff
@@ -0,0 +1,260 @@
+// Copyright Amazon.com, Inc. or its affiliates.
+
+//! Types representing positions, spans, locations, etc of parsed PartiQL text.
+
+use std::fmt;
+use std::fmt::Formatter;
+use std::num::NonZeroUsize;
+use std::ops::{Add, Sub};
+
+macro_rules! impl_pos {
+    ($pos_type:ident, $primitive:ty) => {
+        impl Add for $pos_type {
+            type Output = Self;
+
+            fn add(self, rhs: Self) -> Self::Output {
+                Self(self.0 + rhs.0)
+            }
+        }
+        impl Add<$primitive> for $pos_type {
+            type Output = Self;
+
+            fn add(self, rhs: $primitive) -> Self::Output {
+                Self(self.0 + rhs)
+            }
+        }
+        impl Sub for $pos_type {
+            type Output = Self;
+
+            fn sub(self, rhs: Self) -> Self::Output {
+                Self(self.0 - rhs.0)
+            }
+        }
+        impl Sub<$primitive> for $pos_type {
+            type Output = Self;
+
+            fn sub(self, rhs: $primitive) -> Self::Output {
+                Self(self.0 - rhs)
+            }
+        }
+        impl $pos_type {
+            #[inline(always)]
+            pub fn from_usize(n: usize) -> Self {
+                Self(n as $primitive)
+            }
+
+            #[inline(always)]
+            pub fn to_usize(&self) -> usize {
+                self.0 as usize
+            }
+        }
+        impl From<usize> for $pos_type {
+            fn from(n: usize) -> Self {
+                Self::from_usize(n)
+            }
+        }
+    };
+}
+
+/// A 0-indexed byte offset, relative to some other position.
+///
+/// This type is small (u16 currently) to allow it to be included in ASTs and other
+/// data structures.
+#[derive(Debug, Copy, Clone, Default, PartialEq, Eq, Ord, PartialOrd, Hash)]
+pub struct ByteOffset(pub u16);
+impl_pos!(ByteOffset, u16);
+
+/// A 0-indexed line offset, relative to some other position.
+#[derive(Debug, Copy, Clone, Default, PartialEq, Eq, Ord, PartialOrd, Hash)]
+pub struct LineOffset(pub u16);
+impl_pos!(LineOffset, u16);
+
+/// A 0-indexed char offset, relative to some other position.
+///
+/// This value represents the number of unicode codepoints seen, so will differ
+/// from [`ByteOffset`] for a given location in a &str if the string contains
+/// non-ASCII unicode characters
+#[derive(Debug, Copy, Clone, Default, PartialEq, Eq, Ord, PartialOrd, Hash)]
+pub struct CharOffset(pub u16);
+impl_pos!(CharOffset, u16);
+
+/// A 0-indexed byte absolute position (i.e., relative to the start of a &str)
+///
+/// This type is small (u16 currently) to allow it to be included in ASTs and other
+/// data structures.
```

Comment:
Potentially it might require addition or removal of casts if they were to do calculations on the offsets/positions exposed by the AST. For now, I'll increase it to u32...